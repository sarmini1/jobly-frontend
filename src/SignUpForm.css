.SignUpForm {
    padding: 5% 5% 5% 5%;
    text-align: left;
}

.btn {
    background-color: #a44a3f;
    margin: 5px;
    font-family: 'Comfortaa', cursive;
    color: white;
}

h3 {
    color: #a44a3f;
    font-family: 'Comfortaa', cursive;
}

label {
    color: #a44a3f;
    font-family: 'Comfortaa', cursive;
    line-height: 2px;
}

input {
    margin-bottom: 15px;
}

