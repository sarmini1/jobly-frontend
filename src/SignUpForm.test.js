import React from 'react';
import { render, unmountComponentAtNode } from "react-dom";
import { act } from 'react-dom/test-utils';
import { MemoryRouter } from "react-router-dom";
import SignUpForm from './SignUpForm';

import { fireEvent } from '@testing-library/react';


let container = null;

beforeEach(() => {
  // setup a DOM element as a render target
  container = document.createElement("div");
  // console.log(document);
  document.body.appendChild(container);
});

afterEach(() => {
  // cleanup on exiting
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

it('mounts without crashing', async function () {
  await act(async () => {
    render(
      <MemoryRouter>
        <SignUpForm signUp={() => { }} />
      </MemoryRouter>
      , container);
  });
});

it("matches snapshot", async function () {
  await act(async () => {
    render(
      <MemoryRouter>
        <SignUpForm signUp={() => { }} />
      </MemoryRouter>
      , container);
  });
  expect(container).toMatchSnapshot();
});

it("renders empty form successfully", async function () {

  // Use the asynchronous version of act to apply resolved promises
  await act(async () => {
    render(
      <MemoryRouter>
        <SignUpForm signUp={() => { }} />
      </MemoryRouter>
      , container);
  });

  const form = container.querySelector("form");
  const usernameInput = container.querySelector("#signUpForm-username");
  const passwordInput = container.querySelector("#signUpForm-password");
  const firstNameInput = container.querySelector("#signUpForm-first-name");
  const lasttNameInput = container.querySelector("#signUpForm-last-name");
  const emailInput = container.querySelector("#signUpForm-email");

  expect(container).toContainHTML("<h3>Sign Up</h3>");
  expect(container).toContainElement(form);
  expect(container).toContainElement(usernameInput);
  expect(container).toContainElement(passwordInput);
  expect(container).toContainElement(firstNameInput);
  expect(container).toContainElement(lasttNameInput);
  expect(container).toContainElement(emailInput);
});

it("calls prop function upon submission", async function () {
  const mockSignUp = jest.fn();

  await act(async () => {
    render(
      <MemoryRouter>
        <SignUpForm signUp={mockSignUp} />
      </MemoryRouter>
      , container);
  });

  const form = container.querySelector("form");
  const usernameInput = container.querySelector("#signUpForm-username");
  const passwordInput = container.querySelector("#signUpForm-password");
  const firstNameInput = container.querySelector("#signUpForm-first-name");
  const lasttNameInput = container.querySelector("#signUpForm-last-name");
  const emailInput = container.querySelector("#signUpForm-email");

  await act(async () => {
    fireEvent.change(usernameInput, { target: { value: "test-username" } });
    fireEvent.change(passwordInput, { target: { value: "test-password" } });
    fireEvent.change(firstNameInput, { target: { value: "test-first-name" } });
    fireEvent.change(lasttNameInput, { target: { value: "test-last-name" } });
    fireEvent.change(emailInput, { target: { value: "test-email" } });
    fireEvent.submit(form);
  });

  expect(mockSignUp).toHaveBeenCalledWith(
    {
      username: "test-username",
      password: "test-password",
      firstName: "test-first-name",
      lastName: "test-last-name",
      email: "test-email"
    }
  );
  expect(mockSignUp).toBeCalledTimes(1);

  mockSignUp.mockRestore();
});

it("shows errors when submission function throws them", async function () {
  const mockSignUp = jest
    .fn()
    .mockImplementationOnce(() => Promise.reject(["mock error message"]));

  await act(async () => {
    render(
      <MemoryRouter>
        <SignUpForm signUp={mockSignUp} />
      </MemoryRouter>
      , container);
  });

  const form = container.querySelector("form");
  const usernameInput = container.querySelector("#signUpForm-username");
  const passwordInput = container.querySelector("#signUpForm-password");
  const firstNameInput = container.querySelector("#signUpForm-first-name");
  const lasttNameInput = container.querySelector("#signUpForm-last-name");
  const emailInput = container.querySelector("#signUpForm-email");

  await act(async () => {
    fireEvent.change(usernameInput, { target: { value: "test-username" } });
    fireEvent.change(passwordInput, { target: { value: "test-password" } });
    fireEvent.change(firstNameInput, { target: { value: "test-first-name" } });
    fireEvent.change(lasttNameInput, { target: { value: "test-last-name" } });
    fireEvent.change(emailInput, { target: { value: "test-email" } });
    fireEvent.submit(form);
  });

  expect(mockSignUp).toHaveBeenCalledWith(
    {
      username: "test-username",
      password: "test-password",
      firstName: "test-first-name",
      lastName: "test-last-name",
      email: "test-email"
    }
  );
  expect(mockSignUp).toBeCalledTimes(1);

  // shows errors
  const error = container.querySelector(".Error");
  expect(error).toContainHTML("mock error message");
  expect(container).toContainElement(error);

  // holds onto initial inputs in form
  expect(usernameInput).toContainHTML("test-username");

  mockSignUp.mockRestore();
});

it("can resubmit form when errors are present", async function () {
  const mockSignUp = jest
    .fn()
    .mockImplementationOnce(() => Promise.reject(["mock error message"]));

  await act(async () => {
    render(
      <MemoryRouter>
        <SignUpForm signUp={mockSignUp} />
      </MemoryRouter>
      , container);
  });

  const form = container.querySelector("form");
  const usernameInput = container.querySelector("#signUpForm-username");
  const passwordInput = container.querySelector("#signUpForm-password");
  const firstNameInput = container.querySelector("#signUpForm-first-name");
  const lasttNameInput = container.querySelector("#signUpForm-last-name");
  const emailInput = container.querySelector("#signUpForm-email");

  await act(async () => {
    fireEvent.change(usernameInput, { target: { value: "test-username" } });
    fireEvent.change(passwordInput, { target: { value: "test-password" } });
    fireEvent.change(firstNameInput, { target: { value: "test-first-name" } });
    fireEvent.change(lasttNameInput, { target: { value: "test-last-name" } });
    fireEvent.change(emailInput, { target: { value: "test-email" } });
    fireEvent.submit(form);
  });

  // shows errors
  const error = container.querySelector(".Error");
  expect(error).toContainHTML("mock error message");
  expect(container).toContainElement(error);

  // update some inputs and resubmit form
  await act(async () => {
    fireEvent.change(usernameInput, { target: { value: "new-username2" } });
    fireEvent.change(passwordInput, { target: { value: "new-password2" } });
    fireEvent.submit(form);
  });

  expect(mockSignUp).toHaveBeenCalledWith(
    {
      username: "new-username2",
      password: "new-password2",
      firstName: "test-first-name",
      lastName: "test-last-name",
      email: "test-email"
    }
  );
  expect(mockSignUp).toBeCalledTimes(2);

  mockSignUp.mockRestore();
});
