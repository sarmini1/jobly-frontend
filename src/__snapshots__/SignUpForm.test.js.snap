// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div>
  <div
    class="SignUpForm"
  >
    <div
      class="row"
    >
      <div
        class="col-1 col-sm-2 col-lg-4"
      />
      <div
        class="col-10 col-sm-8 col-lg-4"
      >
        <h3>
          Sign Up
        </h3>
        <form
          class=""
        >
          <label
            class="form-label"
            for="signUpForm-first-name"
          >
            First Name
          </label>
          <input
            autocomplete="first-name"
            class="form-control"
            id="signUpForm-first-name"
            name="firstName"
            required=""
            type="text"
            value=""
          />
          <label
            class="form-label"
            for="signUpForm-last-name"
          >
            Last Name
          </label>
          <input
            autocomplete="last-name"
            class="form-control"
            id="signUpForm-last-name"
            name="lastName"
            required=""
            type="text"
            value=""
          />
          <label
            class="form-label"
            for="signUpForm-email"
          >
            Email
          </label>
          <input
            autocomplete="email"
            class="form-control"
            id="signUpForm-email"
            name="email"
            required=""
            type="text"
            value=""
          />
          <label
            class="form-label"
            for="signUpForm-username"
          >
            Username
          </label>
          <input
            autocomplete="username"
            class="form-control"
            id="signUpForm-username"
            name="username"
            required=""
            type="text"
            value=""
          />
          <label
            class="form-label"
            for="signUpForm-password"
          >
            Password
          </label>
          <input
            autocomplete="current-password"
            class="form-control"
            id="signUpForm-password"
            name="password"
            required=""
            type="password"
            value=""
          />
          <button
            class="btn"
            type="submit"
          >
            Sign Up!
          </button>
        </form>
      </div>
      <div
        class="col-1 col-sm-2 col-lg-4"
      />
    </div>
  </div>
</div>
`;
